lib/Monads/OptionMonadWP.thy

```diff
@@ -167,67 +188,114 @@ lemma owhile_NF [wp]:
 
-lemma ovalidNF_wp_comb3 [wp_comb]:
+lemma ovalidNF_wp_comb3[wp_comb]:
   "\<lbrakk> ovalidNF P f Q; ovalidNF P' f Q' \<rbrakk> \<Longrightarrow> ovalidNF (\<lambda>s. P s \<and> P' s) f (\<lambda>r s. Q r s \<and> Q' r s)"
   by (simp add: ovalidNF_def)
 
 
 
 (* FIXME: WP rules for no_ofail, which might not be correct. *)
-lemma no_ofail_ofail [wp]: "no_ofail (\<lambda>_. False) ofail"
-  by (simp add: no_ofail_def ofail_def)
+lemma no_ofailD:
+  "\<lbrakk> no_ofail P m; P s \<rbrakk> \<Longrightarrow> \<exists>y. m s = Some y"
+  by (simp add: no_ofail_def)
+
+lemma no_ofail_obind2[simp]:
+  assumes f: "no_ofail P f"
+  assumes v: "ovalid Q f R"
+  assumes g: "\<And>r. no_ofail (R r) (g r)"
+  shows "no_ofail (P and Q) (f |>> g)"
+  using v g
+  by (fastforce simp: no_ofail_def obind_def pred_conj_def ovalid_def dest: no_ofailD [OF f])
 
-lemma no_ofail_ogets [wp]: "no_ofail (\<lambda>_. True) (ogets f)"
-  by (simp add: no_ofail_def ogets_def)
+lemma no_ofail_ofail[wp]:
+  "no_ofail \<bottom> ofail"
+  by (simp add: no_ofail_def)
+
+lemma no_ofail_asks_simp[simp]:
+  "no_ofail P (asks f)"
+  unfolding asks_def get_def oreturn_def obind_def no_ofail_def
+  by simp
+
+lemma no_ofail_asks[wp]:
+  "no_ofail \<top> (asks f)"
+  by simp
+
+lemma no_ofail_ogets[wp]:
+  "no_ofail \<top> (ogets f)"
+  by simp
 
-lemma no_ofail_obind [wp]:
+lemma no_ofail_obind[wp]:
   "\<lbrakk> \<And>r. no_ofail (P r) (g r); no_ofail Q f; ovalid Q f P \<rbrakk> \<Longrightarrow> no_ofail Q (obind f g)"
   by (auto simp: no_ofail_def obind_def ovalid_def)
 
-lemma no_ofail_K_bind [wp]:
+lemma no_ofail_K_bind[wp]:
   "no_ofail P f \<Longrightarrow> no_ofail P (K_bind f x)"
   by simp
 
-lemma no_ofail_oguard [wp]:
+lemma no_ofail_oguard[wp]:
   "no_ofail (\<lambda>s. f s) (oguard f)"
   by (auto simp: no_ofail_def oguard_def)
 
-lemma no_ofail_ocondition [wp]:
+lemma no_ofail_ocondition[wp]:
   "\<lbrakk> no_ofail L l; no_ofail R r \<rbrakk>
      \<Longrightarrow> no_ofail (\<lambda>s. if C s then L s else R s) (ocondition C l r)"
   by (simp add: no_ofail_def ocondition_def)
 
-lemma no_ofail_oreturn [wp]:
+lemma no_ofail_oreturn[wp]:
   "no_ofail (\<lambda>_. True) (oreturn x)"
   by (simp add: no_ofail_def oreturn_def)
 
-lemma no_ofail_oskip [wp]:
+lemma no_ofail_oskip[wp]:
   "no_ofail (\<lambda>_. True) oskip"
   by (simp add: no_ofail_def oskip_def)
 
-lemma no_ofail_is_triple [wp_trip]:
+lemma no_ofail_oassert_opt[simp, wp]:
+  "no_ofail (\<lambda>_. P \<noteq> None) (oassert_opt P)"
+  by (simp add: no_ofail_def oassert_opt_def split: option.splits)
+
+lemma no_ofail_owhen[wp]:
+  "(P \<Longrightarrow> no_ofail Q f) \<Longrightarrow> no_ofail (if P then Q else \<top>) (owhen P f)"
+  by (simp add: no_ofail_def owhen_def)
+
+lemma no_ofail_ounless [wp]:
+  "(\<not>P \<Longrightarrow> no_ofail Q f) \<Longrightarrow> no_ofail (if P then \<top> else Q) (ounless P f)"
+  by (simp add: no_ofail_def ounless_def)
+
+lemma no_ofail_oassert [simp, wp]:
```

Comment:
if you are doing a style improvement run, the style improvement run should also extend to the lemmas you're pasting in, no? at least consistently removing the spaces before `[...]` annotations would be nice.